packages/@uppy/aws-s3-multipart/src/utils.ts

```diff
@@ -0,0 +1,28 @@
+import { createAbortError } from '@uppy/utils/lib/AbortController'
+import type { Body as _Body } from '@uppy/utils/lib/UppyFile'
+
+import type { AwsS3Part } from './index'
+
+export function throwIfAborted(signal?: AbortSignal | null): void {
+  if (signal?.aborted) {
+    throw createAbortError('The operation was aborted', {
+      cause: signal.reason,
+    })
+  }
+}
+
+export type UploadResult = { key: string; uploadId: string }
+export type UploadResultWithSignal = UploadResult & { signal?: AbortSignal }
+export type MultipartUploadResult = UploadResult & { parts: AwsS3Part[] }
+export type MultipartUploadResultWithSignal = MultipartUploadResult & {
+  signal?: AbortSignal
+}
+
+export type UploadPartBytesResult = {
+  ETag: string
+  location?: string
+}
+
+export interface Body extends _Body {
+  location: string
+}
```

Comment:
How does this work for the user facing type? Ideally, we have this:

```ts
import Transloadit, { type TransloaditBody } from `@uppy/transloadit`
import Uppy from '@uppy/core'

type Meta = { foo: number }

new Uppy<Meta, TransloaditBody>()
```

Does this mean people have to pass this conditionally based on which method they use?

Reply:
> Does this mean people have to pass this conditionally based on which method they use?

What do you mean? Like if the user overrides the method to upload to a non-Companion instance that do not respond with `{ "location": "…" }`? That would be a tough one to support type-wise for sure, but I think it still quite fuzzy because we're in the middle of the transition, maybe we should wait to be done with it to assess whether we can improve the UX.